File: src/Parrot/Parser/Stream.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Parrot.Lexer;

namespace Parrot.Parser
{
    /// <summary>
    /// TODO: Update summary.
    /// </summary>
    public class Stream
    {
        private readonly IList<Token> _list;
        private int _index;
        private readonly int _count;

        public Stream(IList<Token> source)
        {
            _list = source;
            _index = -1;
            _count = source.Count;
        }

        public Token Peek()
        {
            var temp = _index + 1;
            while (temp < _count)
            {
                if (_list[temp].Type != TokenType.Whitespace)
                {
                    return _list[temp];
                }

                temp++;
            }

            return null;
        }

        public void NextNoReturn()
        {
            GetNextNoReturn();
        }

        public void GetNextNoReturn()
        {
            _index++;
            while (_index < _count && _list[_index].Type == TokenType.Whitespace)
            {
                _index++;
            }
        }

        public Token Next()
        {
            _index ++;
            while (_index < _count)
            {
                if (_list[_index].Type != TokenType.Whitespace)
                {
                    return _list[_index];
                }

                _index++;
            }

            return null;
        }
    }
}

File: src/Parrot.Javascript/scripts/Parser/stream.ts
/// <reference path="../lexer/token.ts"/>
/// <reference path="../lexer/tokenType.ts"/>
class Stream {
    private _list: Token[];
    private _index: number;
    private _count: number;

    constructor(source: Token[]) { 
        this._list = source;
        this._count = source.length;
        this._index = -1;
    }

    peek() : Token {
        var temp = this._index + 1;
        while (temp < this._count) {
            if (this._list[temp].type != TokenType.whitespace) {
                return this._list[temp];
            }

            temp++;
        }

        return null;
    }
    
    getNextNoReturn() {
        this._index++;
        while (this._index < this._count && this._list[this._index].type == TokenType.whitespace) {
            this._index++;
        }
    }

    nextNoReturn() {
        this.getNextNoReturn();
    }

    next() : Token {
        this._index++;
        while (this._index < this._count) {
            if (this._list[this._index].type != TokenType.whitespace) {
                return this._list[this._index];
            }

            this._index++;
        }

        return null;
    }
}